jaxquantum/devices/superconducting/transmon.py

```diff
@@ -65,35 +68,62 @@ def common_ops(self):
-            )
+            n_array = jnp.arange(-n_max, n_max + 1)
+            ops["n"] = jnp2jqt(jnp.diag(n_array))
 
-        elif self.basis == BasisTypes.single_charge:
+            n_minus_ng_array = n_array - self.params["ng"] * jnp.ones(N)
+            ops["H_charge"] = jnp2jqt(jnp.diag(4 * self.params["Ec"] * n_minus_ng_array**2))
+
+        elif self.basis == BasisTypes.singlecharge:
             """
             Here H = Ec (n - 2ng)² - Ej cos(φ) in the single-electron charge basis. Using Eq. (5.36) of Kyle Serniak's
             thesis, we have H = Ec ∑ₙ(n - 2*ng) |n⟩⟨n| - Ej/2 * ∑ₙ|n⟩⟨n+2| + h.c where n counts the number of electrons, 
             not Cooper pairs. Note, we use 2ng instead of ng to match the gate offset charge convention of the transmon 
             (as done in Kyle's thesis).
             """
-            n_max = (N - 1) // 2
+            n_max = (N) // 2
 
             ops["id"] = identity(N)
             ops["cos(φ)"] = 0.5 * (jnp2jqt(jnp.eye(N, k=2) + jnp.eye(N, k=-2)))
             ops["sin(φ)"] = 0.5j * (jnp2jqt(jnp.eye(N, k=2) - jnp.eye(N, k=-2)))
             ops["cos(φ/2)"] = 0.5 * (jnp2jqt(jnp.eye(N, k=1) + jnp.eye(N, k=-1)))
             ops["sin(φ/2)"] = 0.5j * (jnp2jqt(jnp.eye(N, k=1) - jnp.eye(N, k=-1)))
-            ops["n"] = jnp2jqt(jnp.diag(jnp.arange(-n_max, n_max + 1)))
 
-            n_minus_ng_array = jnp.arange(-n_max, n_max + 1) - 2 * self.params[
-                "ng"
-            ] * jnp.ones(N)
+            n_array = jnp.arange(-n_max, n_max)
+            ops["n"] = jnp2jqt(jnp.diag(n_array))
+            n_minus_ng_array = n_array - 2 * self.params["ng"] * jnp.ones(N)
+            ops["H_charge"] = jnp2jqt(jnp.diag(self.params["Ec"] * n_minus_ng_array**2))
+
+        elif self.basis == BasisTypes.singlecharge_even:
+            n_max = N
+
+            ops["id"] = identity(n_max)
+            ops["cos(φ)"] = 0.5 * (jnp2jqt(jnp.eye(n_max, k=1) + jnp.eye(n_max, k=-1)))
+            ops["sin(φ)"] = 0.5j * (jnp2jqt(jnp.eye(n_max, k=1) - jnp.eye(n_max, k=-1)))
+            ops["cos(2φ)"] = 0.5 * (jnp2jqt(jnp.eye(n_max, k=2) + jnp.eye(n_max, k=-2)))
+            ops["sin(2φ)"] = 0.5j * (jnp2jqt(jnp.eye(n_max, k=2) - jnp.eye(n_max, k=-2)))
+
+            n_array = jnp.arange(-n_max, n_max, 2)
+            ops["n"] = jnp2jqt(jnp.diag(n_array))
+            n_minus_ng_array = n_array - 2 * self.params["ng"] * jnp.ones(n_max)
+            ops["H_charge"] = jnp2jqt(jnp.diag(self.params["Ec"] * n_minus_ng_array**2))
+
+        elif self.basis == BasisTypes.singlecharge_odd:
+            n_max = N
+
+            ops["id"] = identity(n_max)
+            ops["cos(φ)"] = 0.5 * (jnp2jqt(jnp.eye(n_max, k=1) + jnp.eye(n_max, k=-1)))
+            ops["sin(φ)"] = 0.5j * (jnp2jqt(jnp.eye(n_max, k=1) - jnp.eye(n_max, k=-1)))
+            ops["cos(2φ)"] = 0.5 * (jnp2jqt(jnp.eye(n_max, k=2) + jnp.eye(n_max, k=-2)))
+            ops["sin(2φ)"] = 0.5j * (jnp2jqt(jnp.eye(n_max, k=2) - jnp.eye(n_max, k=-2)))
```

Comment:
Probably worth not repeating code as much as possible to minimize possibility of mistakes now or in future updates.